src/wiseagents/wise_agent_impl.py

```diff
@@ -632,6 +636,316 @@ def set_response_delivery(self, response_delivery: Callable[[], WiseAgentMessage
+
+        # Generate a chat ID that will be used to collaborate on this query
+        chat_id = str(uuid.uuid4())
+
+        ctx = WiseAgentRegistry.get_or_create_context(request.context_name)
+
+        # Determine the agents required to answer the query
+        agent_selection_prompt = ("Given the following query and a description of the agents that are available," +
+                                  " determine all of the agents that could be required to solve the query." +
+                                  " Format the response as a space separated list of agent names and don't include " +
+                                  " anything else in the response.\n" +
+                                  " Query: " + request.message + "\n" + "Available agents:\n" +
+                                  "\n".join(WiseAgentRegistry.get_agent_names_and_descriptions()) + "\n")
+        ctx.append_chat_completion(chat_uuid=chat_id, messages={"role": "system", "content": self.llm.system_message})
+        ctx.append_chat_completion(chat_uuid=chat_id, messages={"role": "user", "content": agent_selection_prompt})
+
+        logging.debug(f"messages: {ctx.llm_chat_completion[chat_id]}")
+        llm_response = self.llm.process_chat_completion(ctx.llm_chat_completion[chat_id], tools=[])
+        ctx.append_chat_completion(chat_uuid=chat_id, messages=llm_response.choices[0].message)
+
+        # Assign the agents to phases
+        agent_assignment_prompt = ("Assign each of the agents that will be required to solve the query to one of the following phases:\n" +
+                                   ", ".join(self.phases) + "\n" +
+                                   "Assume that agents within a phase will be executed in parallel." +
+                                   " Format the response as a space separated list of agents for each phase, where the first"
+                                   " line contains the list of agents for the first phase and second line contains the list of"
+                                   " agents for the second phase and so on. Don't include anything else in the response.\n")
+        ctx.append_chat_completion(chat_uuid=chat_id, messages={"role": "user", "content": agent_assignment_prompt})
+        llm_response = self.llm.process_chat_completion(ctx.llm_chat_completion[chat_id], tools=[])
+        ctx.append_chat_completion(chat_uuid=chat_id, messages=llm_response.choices[0].message)     
+        phases = [phase.split() for phase in llm_response.choices[0].message.content.splitlines()]
+        ctx.set_agent_phase_assignments(chat_id, phases)
+        ctx.set_current_phase(chat_id, 0)
+        ctx.add_query(chat_id, request.message)
+
+        # Kick off the first phase
+        for agent in phases[0]:
+            self.send_request(WiseAgentMessage(message=request.message, sender=self.name,
+                                               context_name=request.context_name, chat_id=chat_id), agent)
+
```

Comment:
So if I well understand here you send the original request to each agent for the current phase. Do you plan to split the original request.message somehow to ask more preciase question to each agents? 
Maybe something for which would make sense start a github discussion?

Reply:
Will start a discussion about this. We could probably use the agent's description to create a more precise question for each agent.